Latex rendering with react-native-math-view

KobeBergmans opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A brief description of what this pull request does.
-->
Added rendering of latex code block with the react-native-math-view library. Latex code is split up in different lines and these are rendered independently. By default the code block displays 2 lines of latex code. If you click on the block the next screen will display all the code in a horizontal and vertical scroll view.

To get the inline functionality working I changed the commonmark and commonmark-react-renderer library. The PR's for those library's can be found here:
- https://github.com/mattermost/commonmark.js/pull/11
- https://github.com/mattermost/commonmark-react-renderer/pull/3

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-43469

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates

#### Device Information
This PR was tested on: 
- Pixel 5: Android studio emulator
- OnePlus Nord 2 5G: Oxygen OS V11.3
- Iphone 12: Xcode emulator

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs/Videos (for both iOS and Android if possible).
-->
<p float="left">
<img src="https://user-images.githubusercontent.com/55985776/131311854-235547e8-b2df-4559-902e-6acc7233879e.jpg" width=350>
<img src="https://user-images.githubusercontent.com/55985776/131311861-690661f0-ebe3-4dd2-aaed-0cf47ac86bac.jpg" width=350>
</p>
<p float="left">
<img src="https://user-images.githubusercontent.com/55985776/131311865-a18e809d-4dcd-47b0-9b96-9821c913632c.jpg" width=350>
<img src="https://user-images.githubusercontent.com/55985776/131311875-c9e42bcc-b2fc-4faf-9f10-156011d84d79.png" width=350>
</p>
<p float="left">
<img src="https://user-images.githubusercontent.com/55985776/131311895-3aafb384-906b-4c2c-ac4e-1a995ea984f7.png" width=350>
<img src="https://user-images.githubusercontent.com/55985776/131311900-705123fb-d3a9-4a03-8e9a-220c2f3eb66c.png" width=350>
</p>



#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note
Added rendering of latex code blocks. Added latex screen.
```

esethna reviewed (commented):
@KobeBergmans thanks! Does this support the same in-line syntax as the webapp PR you submitted? Similarly, do the config settings we discussed on the webapp PRs also apply here?

KobeBergmans commented:
@esethna This PR does not include inline rendering of latex it just renders the usual latex code block that was present before my previous PR. I am currently trying to find a solution for the inline latex rendering too but that has not yet worked nicely.

KobeBergmans commented:
I made the inline latex work like this: 
![Inline code](https://user-images.githubusercontent.com/55985776/131513626-5cd451c3-fdb6-4b82-b792-8bca5139f18e.png)

The code does not work 100% correct yet but does this seem like a good enough solution to be implemented?

esethna commented:
Nice, thanks @KobeBergmans! I'll let @enahum comment on the code/approach.

KobeBergmans commented:
@enahum @anurag6713 I polished the inline latex functionality and added some tests for it. I merged the code into this PR normally.

KobeBergmans commented:
@enahum, @anurag6713 I updated the latex rendering to also not trigger when the `$` is preceded by a word character as suggested here: https://github.com/mattermost/mattermost-webapp/pull/8644#issuecomment-918335216

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

KobeBergmans commented:
@esethna Is someone still looking in to this? Because I haven't had any feedback from anyone...

enahum commented on app/components/markdown/markdown.js:
```diff
@@ -169,6 +179,41 @@ export default class Markdown extends PureComponent {
+                    <Text
+                        style={{lineHeight: this.state.inlineLatexHeight}}
+                    >
+                        {this.renderText({context, literal: firstPart})}
```
this is calling itself, is that expected?

Also, perhaps you can extract all this if block to it's own function for readability.

KobeBergmans replied:
That behaviour is expected. What this function does is check if there is an inline latex match. If there is a match it splits the text up in text and latex code. It calls itself on the text part to check if there are maybe more inline latex code blocks and renders the latex code.

I'll split this function up.

enahum commented on app/components/markdown/markdown.js:
```diff
@@ -169,6 +179,41 @@ export default class Markdown extends PureComponent {
+                        {this.renderText({context, literal: firstPart})}
+                        <LatexInline
+                            content={latexCode}
+                            onLayout={(event) => {
```
do not use inline functions

enahum commented on app/screens/latex/latex.js:
```diff
@@ -0,0 +1,113 @@
+        content: PropTypes.string.isRequired,
+    };
+
+    componentDidMount() {
+        BackHandler.addEventListener('hardwareBackPress', this.handleAndroidBack);
+    }
+
+    componentWillUnmount() {
+        BackHandler.removeEventListener('hardwareBackPress', this.handleAndroidBack);
+    }
+
+    handleAndroidBack = () => {
+        popTopScreen();
+        return true;
+    };
```
Any reason why this is needed? is it not handled by default?

KobeBergmans replied:
I copied this from the code screen but i think it might be a better solution to just extend the Latex screen from the code screen. I'll implement that so there is no duplicate code

enahum commented on app/screens/latex/latex.js:
```diff
@@ -0,0 +1,113 @@
+
+        return (
+            <SafeAreaView
+                edges={['bottom', 'left', 'right']}
```
why is top is being omitted?

KobeBergmans replied:
I don't know why the top is omitted but i based my code on the `Code` screen and there it is omitted too.

enahum commented on app/screens/latex/latex.js:
```diff
@@ -0,0 +1,113 @@
+                            >
+                                <MathView
+                                    math={latexCode}
+                                    onError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Error: ' + errorMsg.message}</Text>;
+                                    }}
+                                    renderError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Render error: ' + errorMsg.error.message}</Text>;
+                                    }}
```
no inline functions

enahum commented on app/screens/latex/latex.js:
```diff
@@ -0,0 +1,113 @@
+        errorText: {
+            fontSize: 14,
+            marginHorizontal: 5,
+            color: 'rgb(255, 0, 0)',
```
errorTextColor

enahum commented on ios/Gemfile.lock:
```diff
@@ -57,6 +57,7 @@ GEM
     ethon (0.12.0)
       ffi (>= 1.3.0)
     ffi (1.15.0)
+    ffi (1.15.0-x64-mingw32)
```
why was this added?

KobeBergmans replied:
This must have been an accident while trying to get my workspace up and running. I'll delete this

enahum commented on jest.config.js:
```diff
@@ -28,6 +28,6 @@ module.exports = {
     },
     transformIgnorePatterns: [
-        'node_modules/(?!(react-native|@react-native|jail-monkey|@sentry/react-native|react-navigation|@react-native-community/cameraroll))',
+        'node_modules/(?!(react-native|@react-native|jail-monkey|@sentry/react-native|react-navigation|@react-native-community/cameraroll|hast-util-from-selector|hastscript|property-information|hast-util-parse-selector|space-separated-tokens|comma-separated-tokens|zwitch))',
```
can you explain why this changes are required ?

KobeBergmans replied:
This happens when a module is not transformed by the typescript compiler. Jest gives following error:

```
    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/en/ecmascript-modules for how to enable it.
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:
    export function zwitch(key, options) {
    ^^^^^^

    SyntaxError: Unexpected token 'export'

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1350:14)
      at Object.<anonymous> (node_modules/hast-util-from-selector/index.js:22:1)
```

So the modules need to be forced to be transformed by jest.

You can read more on this in following thread: https://stackoverflow.com/questions/49263429/jest-gives-an-error-syntaxerror-unexpected-token-export

hmhealey commented on app/components/markdown/latex_code_block/latex_code_block.js:
```diff
@@ -0,0 +1,204 @@
+                            >
+                                <MathView
+                                    math={latexCode}
+                                    onError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Error: ' + errorMsg.message}</Text>;
+                                    }}
+                                    renderError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Render error: ' + errorMsg.error.message}</Text>;
+                                    }}
```
^ The current code makes it so that the MathView will re-render whenever the LatexCodeBlock does because it's passing a new copy of the arrow function in each time. If they're methods on the LatexCodeBlock, it'll always pass the a reference to the same method.

KobeBergmans replied:
Ah ok i didn't know that. I'll change that to a class method

KobeBergmans commented on package.json:
```diff
@@ -58,6 +58,7 @@
     "react-native-linear-gradient": "2.5.6",
     "react-native-local-auth": "1.6.0",
     "react-native-localize": "2.1.1",
+    "react-native-math-view": "3.9.4",
```
I started working on this a couple months back and the newest version was not out back then. But i'll see if i can upgrade to the newer version

KobeBergmans commented on app/components/markdown/latex_code_block/latex_code_block.js:
```diff
@@ -0,0 +1,204 @@
+
+export default class LatexCodeBlock extends MarkdownCodeBlock {
+    static defaultProps = {
+        language: 'latex',
```
Seems like a good idea, I just kept it there from the default MarkdownCodeBlock but it is indeed not necessary

KobeBergmans commented on app/components/markdown/latex_code_block/latex_code_block.js:
```diff
@@ -0,0 +1,204 @@
+
+const MAX_LINES = 2;
+
+export default class LatexCodeBlock extends MarkdownCodeBlock {
```
I had a look at changing the files to typescript but it seemed a lot more complex than the webapp because I had no real example to go by. But if I have some time left in the following week I will have a look at that.

In the mean time I'll add the `propTypes`

hmhealey replied:
Works for me. It's been a while since I've worked on the mobile app, and I know most of the TypeScript migration effort has been on the web-side, so I'm not surprised that it's trickier to do here.

KobeBergmans commented on app/components/markdown/latex_code_block/latex_code_block.js:
```diff
@@ -0,0 +1,204 @@
+    });
+
+    trimContent = (content) => {
+        const lines = splitLatexCodeInLines(content);
```
The library I used to implement the latex code does not support the standard Latex linebreaks (syntax: `\\`). So if this function is not called there are no linebreaks in the rendered code

KobeBergmans commented on app/components/markdown/markdown.js:
```diff
@@ -446,7 +501,14 @@ export default class Markdown extends PureComponent {
+        let ast;
+
+        if (this.props.enableLatex && this.props.enableInlineLatex) {
+            const escapedContent = escapeLatexInputPreParser(this.props.value.toString());
```
The way I solved this in the web app PR was by modifying the markdown parser. I made the parser stop at `$` and then check if there was any latex code ahead. If there was latex code ahead it was parsed into the AST as one big block and the different markdown syntax components where ignored (like `_emphasis_`).

I tried to fix this problem by adding a transformation as you described but this didn't work when there were markdown syntax compontents inside of the inline latex code. For example: `$x_0 = x_1 = /cdots$` would not render correctely because the markdown library would parse `0 = x` as a block that would needed to get emphasis. 

It is possible to overcome this issue by manipulating the AST and checking if there is special syntax inside inline latex code as you said. But that seemed a lot more complicated than doing it this way.

hmhealey replied:
Perhaps it would be easier to just modify the Markdown library to do that then. I think I'd feel more comfortable doing that than escaping and unescaping these since the current way of doing this feels like it could be fragile since anything else handled to the Markdown code could interfere with this as it currently is. For example, if at-signs or number signs are used in the Latex, I think that'll cause the Markdown parser to interfere since the mobile app's version is actually responsible for doing at-mentions and hashtags as well.

Overall, I don't think it should be that hard to add since [our fork of commonmark.js](https://github.com/mattermost/commonmark.js) is much better maintained, and I've generally found it easier to work with than marked, but it does require changes to two libraries to change: that fork and https://github.com/mattermost/commonmark-react-renderer.

You'll have to make most of your changes in [`inlines.js`](https://github.com/mattermost/commonmark.js/blob/master/lib/inlines.js) by adding an extra case to `parseInlines`. It's a bit more powerful than the `marked` version, so you can backtrack to check what's before the $ without having to use a lookbehind (see the at mention or hashtag parsing for an example of that).

Then, you'll want to add a new `Node` type for the inline latex and ensure commonmark-react-renderer has some basic handling for it since it needs to know about every potential type of element for some reason that I can't quite remember.

Apologies for making this more complicated, but I think avoiding the escaping logic will be safer in the long run.

KobeBergmans replied:
I understand that this way of doing things is indeed a bit safer and more robust so I'll take a look at this in the following week.

I thought about changing the commonmark library when I started working on this PR but I looked at the packages and it seemed like the app used the standard commonmark packages. Then I looked into the patches that are done after the npm install process and I guess that is where the changes occur.

So how do I change the commonmark library's? Do I just make a PR on the mattermost fork and you guys will change the patch file or how does this work?

KobeBergmans commented on app/utils/latex.ts:
```diff
@@ -0,0 +1,127 @@
+
+    let i = 0;
+    while (i < latexCode.length) {
+        const firstBegin = latexCode.indexOf('\\begin{', i);
```
Unfortunaltely the react-native framework is not really that compatible with web library's. I looked into using a webView component to render the latex code but that was an ugly solution because inline this did not work well. So the `katex` library was not really an option on mobile

The `react-native-math-view library` is a way to fix this issue, but the drawback is that it does not support line breaks. So they need to be added manually

hmhealey replied:
I see what you mean. This feels like why I ended up writing so much of the mobile app's Markdown code from scratch. Unfortunately, react-native has a lot fewer libraries for things like this, so we end up having to do a lot ourselves.

KobeBergmans commented on app/utils/latex.ts:
```diff
@@ -0,0 +1,127 @@
+        } else if (i === outLines.length - 2) {
+            outLines = outLines.slice(0, i).concat([outLines[i] + '\\\\' + outLines[i + 1]]);
+        } else {
+            return outLines;
```
That is indeed true, I missed that

KobeBergmans commented on app/utils/latex.ts:
```diff
@@ -0,0 +1,127 @@
+ * - The linebreak is in between brackets.
+ * - The linebreak occurs inbetween a `\begin` and `\end` statement.
+ */
+export function splitLatexCodeInLines(content: string): Array<string> {
```
The linebreaks inside of a begin/end statement or curly braces are handled correctely by the `react-native-math-view` library.

For example when creating a matrix:
```
\begin{matrix}
1 & 2 & 3\\
a & b & c
\end{matrix}
```

The matrix still needs to be printed on 1 line because the linebreaks appear inside of the matrix statement. 

The same thing happens when a line break is in between curly braces. An example of this is the following:
```
\prod_{\substack{1\le i \le n\\ 1\le j \le m}}M_{i,j}
```

hmhealey replied:
Ah, I see. So that's needed because the react-native-math-view doesn't know how to handle the line breaks itself when it would better understand how to apply them in different contexts.

Thanks for the examples since I've only got a very limited understanding of Latex.

KobeBergmans commented on app/utils/latex.ts:
```diff
@@ -0,0 +1,127 @@
+    i = 0;
+    while (i < latexCode.length) {
+        const firstBegin = latexCode.indexOf('{', i);
+        const firstEnd = latexCode.indexOf('}', i);
```
I don't know if you wanted to have the double backslash or single so i did both.

The single backslash `\begin{}}\end{}` is not split up.
Double backslash `\\begin{}}\\end{}` is split in 2 lines. A first empty line and a second one that has `begin{}}\\end{}` wich gives a render error

hmhealey replied:
Cool. As long as it errors nicely without crashing the app, that's good with me. Taking another look at this, I can tell that it won't ever loop infinitely, and that's what I was most worried about anyway.

KobeBergmans commented:
@enahum, @hmhealey I think I got to all the requested changes, if I missed one or you have more questions let me know!

hmhealey commented on app/components/markdown/latex_code_block/latex_code_block.js:
```diff
@@ -0,0 +1,209 @@
+            content,
+        };
+
+        const languageDisplayName = getDisplayNameForLanguage(language);
```
To follow up on https://github.com/mattermost/mattermost-mobile/pull/5651#discussion_r718550915, could you remove the language prop from here as well?

KobeBergmans commented:
@enahum @hmhealey I played around with the patch files. I think I did everything right but on my machine the patch does not work, the inline latex features are broken. Can someone maybe generate the right files for me? Because I can't seem to find the mistake.

hmhealey commented:
@enahum Do we have any docs for doing that, or are you able to help generate those?

enahum commented:
it seems like the fork is behind our commonmark fork, perhaps that is the problem

![image](https://user-images.githubusercontent.com/6757047/140623122-cd285d9d-7bb9-465b-8d0a-c9fe62074003.png)

![image](https://user-images.githubusercontent.com/6757047/140623130-fb0aedf4-9655-4236-8e62-0ca1a2d02b04.png)

KobeBergmans commented:
@enahum Thanks for the tip I made the patches from the wrong repository. 

Now everything works as espected with the patches.

hmhealey reviewed (approved):
Cool, the patches look good to me

enahum reviewed (approved):
this LGTM. @Willyfrog can you create a ticket for this to get ported to Gekidou? Thanks!!

enahum commented:
@KobeBergmans can you please resolve the conflicts ?

Willyfrog commented:
> this LGTM. @Willyfrog can you create a ticket for this to get ported to Gekidou? Thanks!!

[done](https://mattermost.atlassian.net/browse/MM-40194)

KobeBergmans commented:
I think i resolved all the conflicts by merging the master branch @enahum

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/19213  
https://circleci.com/gh/mattermost/mattermost-mobile/19212

mattermod commented:
Artifact links:  
https://19213-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.apk  
https://19212-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.ipa

josephbaylon reviewed (changes requested):
@KobeBergmans Here's the result of initial testing for both iOS and Android. Bolded ones are potential issues. Please look into.

A. Latex code preview and opening latex code itself
```
\```latex
X_k = \sum_{n=0}^{2N-1} x_n \cos \left[\frac{\pi}{N} \left(n+\frac{1}{2}+\frac{N}{2}\right) \left(k+\frac{1}{2}\right) \right]
\```
```
1. Works fine on Android
2. Works fine on iOS

B. Inline latex
```
a couple of equations in 1 message: $6^2 - 5 = 36-5 = 31$, another one: $L' = {L}{\sqrt{1-\frac{v^2}{c^2}}}$. Maxwell: $E'=\nabla \times B - 4\pi j$. A hard equation: $\lim_{x\to 0}{\frac{e^x-1}{2x}} \overset{\left[\frac{0}{0}\right]}{\underset{\mathrm{H}}{=}} \lim_{x\to 0}{\frac{e^x}{2}}={\frac{1}{2}}$. Some weird symbols: $y = a + f(\underbrace{b x}_{\ge 0 \text{ by assumption}}) = a + f(\underbrace{b x}_{\mathclap{\ge 0 \text{ by assumption}}})$. Verry long arrows: $A \xleftarrow{\text{this way}} B \xrightarrow[\text{or that way}]{ } C$. Sample integral: $b_n=\frac{1}{\pi}\int\limits_{-\pi}^{\pi}f(x)\sin nx\,\mathrm{d}x=\frac{1}{\pi}\int\limits_{-\pi}^{\pi}x^2\sin nx\,\mathrm{d}x$
```
1. Works fine on Android
2. **Overlaps on iOS (See image)**
![ios_latex_inline](https://user-images.githubusercontent.com/487991/142920352-797e758b-6a6d-40ee-82ef-7ab2394bb243.png)

C. Inline latex triggers
```
1. some text with 15$ and 20$
2. some text with $15 and $20
3. some text with 15$ and $20
4. some text with $20 and 20$
5. [$1+2=3$](https://example.com)
6. _$1+2=3$_
7. **$1+2=3$**
8. ~~$1+2=3$~~
9. https://example.com/$/test/$foo
10.
\```
$1+2=3$
\```
11. `$1+2=3$`
```
1. **On Android, items 5 and 6 do not seem to render properly (See image)**
![android_latex_inline_trigger](https://user-images.githubusercontent.com/487991/142921656-dc65d77b-175f-4727-9d30-5eea70e1fd7a.png)

2. **On iOS, items 5, 6, 7, 8 do not seem to render properly (See image)**
![ios_latex_inline_trigger](https://user-images.githubusercontent.com/487991/142922875-2acd9384-3989-4bc4-98dc-6b11dc0b9abf.png)

D. Inline rainbow
```
Inline-rainbow
# $\color{red}r\color{orange}a\color{yellow}i\color{green}n\color{blue}b\color{Indigo}o\color{violet}w$
```
1. Works fine on Android
2. **Blank on iOS**

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

KobeBergmans commented:
oh I totaly forgot to take a look at this... I'll take a look at it right now.

KobeBergmans commented:
I took a look at the initial testing you did @josephbaylon. I looked into the Android problems in particular because at this time I don't have acces to a mac to test the IOS bugs, I did try to comment on them to help you in further testing. If there is a way I can test and resolve these issues without a mac please point them out and I'll take a look at it.

B, Iphone:
The overlap did happen to me too 1 or 2 times if I was testing on the android emulator. I never got this on my android device though so I figured it was due to the emulator running slow. Did you test this on a real device and did it always happen?

C, 5,6 (both platforms):
I tried to figure out why this was not rendering properly but i could not find out why... It must be that the commonmark library works differently than the marked library (used for the webapp).
I did find out that the problem only occurs for links and emphasis when the `_` symbol is used (with `*` it works fine). So I guess it has to do with `_` and `[]` because it works for the other delimiters? Maybe @hmhealey can help me on this one?

C, 7,8, Iphone:
I fear that this has to do with the react-native-math-view library... I can't find any other reason for the different behaviour on both platforms. So I think the library can't render latex properly on IOS if it is in an environement where there are certain styles applied.
To test this further can you maybe test is `*$1+2$*` renders?

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/19368  
https://circleci.com/gh/mattermost/mattermost-mobile/19367

mattermod commented:
Artifact links:  
https://19368-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.ipa  
https://19367-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.apk

josephbaylon commented:
@KobeBergmans You can maybe use a mac virtual machine to test on iOS simulator.

B, Iphone: I test on real device (iPhone X 15.2) and the overlap always happens
C, 7,8, Iphone: `*$1+2$*` does not render on iPhone but renders on Android

I'll let @hmhealey and @enahum chime in on `C, 5,6 (both platforms)` and `D, Iphone`
https://github.com/mattermost/mattermost-mobile/pull/5651#issuecomment-985403167
https://github.com/mattermost/mattermost-mobile/pull/5651#pullrequestreview-812937264

enahum commented:
C. 5 seems to be a link and I would expect it to render as such
C. 6 to 8 I would expect to render the text as italic, bold, strikethrough on both platforms
D is probably not supported by the new library on iOS (not sure if that is the actual case)

So for me C5 is fine, C6-8 regular markdown should win and render properly and D needs confirmation, as if that is the case we can add it as a limitation in the release notes.

@hmhealey thoughts on this?

hmhealey commented:
Yeah, I agree that C5-8 should prioritize the Markdown over Latex. I think the reason C7 behaves differently is because of [this line](https://github.com/mattermost/commonmark.js/blob/master/lib/inlines.js#L1147) since `_` is a word character in regex. To make it not parse inline latex inside of a link or bold/italic delimiters, I think you can check if `this.brackets` and `this.delimiters` respectively, and if they're both falsey, you should be safe to check for inline latex.

For D, I agree that that one could be a known issue since I think Elias is right that it's a library issue

KobeBergmans commented:
Ok I'll try to take a look at this during this weekend. I'll try to have a go at the mac emulator.

Is it necessary that the webapp and the mobile app have the same functionality because I think `C5-8` do render as inline latex in the webapp so maybe that needs some changing too @hmhealey?

I also came up with the same conclusion on the word character but I changed the regex to not include word characters and it did not work. But I'll take another look at that this weekend.

hmhealey commented:
I think that's fine if they behave differently in those cases even if it's not 100% ideal. Eventually, both the mobile and web app will both be using commonmark.js, so they should behave the same then anyway.

KobeBergmans commented:
> Yeah, I agree that C5-8 should prioritize the Markdown over Latex. I think the reason C7 behaves differently is because of [this line](https://github.com/mattermost/commonmark.js/blob/master/lib/inlines.js#L1147) since `_` is a word character in regex. To make it not parse inline latex inside of a link or bold/italic delimiters, I think you can check if `this.brackets` and `this.delimiters` respectively, and if they're both falsey, you should be safe to check for inline latex.
> 
> For D, I agree that that one could be a known issue since I think Elias is right that it's a library issue

This solution seemed to work for me so I implemented it and added another PR for the commonmark library. so problem C is now solved ;)

I am still working on the mac VM for problem B

KobeBergmans commented:
I finally got the MacOS VM working on my machine, I'll try to get a look at the problem in the folowing week

KobeBergmans commented:
> @KobeBergmans any progress on this?

I am trying to get everything working on the mac vm but it is not that easy it seems. I currently also have exams so I don't have a lot of time to look into this...

esethna commented:
Hey @KobeBergmans, just checking in on this. Please let us know if you have questions or are still working on this.

KobeBergmans commented:
> Hey @KobeBergmans, just checking in on this. Please let us know if you have questions or are still working on this.

Hey @esethna I am still trying to get things working on my Mac VM  to solve the problem on Iphone but I think I am getting there. It is pretty busy for me these days though thus I think it will take some time before it gets resolved

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/21840  
https://circleci.com/gh/mattermost/mattermost-mobile/21839

mattermod commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/f4977c7b-983c-4e37-88c3-cf3fcedcf3d6/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa  
https://output.circle-artifacts.com/output/job/865ce21b-7b52-4a10-9a50-53a5184e1406/artifacts/0/tmp/artifacts/Mattermost_Beta.apk

josephbaylon reviewed (approved):
LGTM. Latex rendering is now working properly on both iOS and Android. Previous issues found are now fixed. Thanks @KobeBergmans !!

enahum commented:
Created a ticket to port this functionality to v2 https://mattermost.atlassian.net/browse/MM-43510

KobeBergmans commented:
No problem @josephbaylon I am terribly sorry for the long wait it should not have taken this long...

enahum commented:
I'm going to manually cherry pick this one as well.
